x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvPercentile.java

```diff
@@ -0,0 +1,602 @@
+        "MvPercentile",
+        MvPercentile::new
+    );
+
+    /**
+     * 2^52 is the smallest integer where it and all smaller integers can be represented exactly as double
+     */
+    private static final double MAX_SAFE_LONG_DOUBLE = Double.longBitsToDouble(0x4330000000000000L);
+
+    private final Expression field;
+    private final Expression percentile;
+
+    @FunctionInfo(
+        returnType = { "double", "integer", "long" },
+        description = "Converts a multivalued field into a single valued field containing "
+            + "the value at which a certain percentage of observed values occur.",
+        examples = @Example(file = "mv_percentile", tag = "example")
+    )
+    public MvPercentile(
+        Source source,
+        @Param(name = "number", type = { "double", "integer", "long" }, description = "Multivalue expression.") Expression field,
+        @Param(
+            name = "percentile",
+            type = { "double", "integer", "long" },
+            description = "The percentile to calculate. Must be a number between 0 and 100. "
+                + "Numbers out of range will return a null instead."
+        ) Expression percentile
+    ) {
+        super(source, List.of(field, percentile));
+        this.field = field;
+        this.percentile = percentile;
+    }
+
+    private MvPercentile(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(field);
+        out.writeNamedWriteable(percentile);
+    }
+
+    @Override
+    protected Expression.TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        return isType(field, dt -> dt.isNumeric() && dt != UNSIGNED_LONG, sourceText(), FIRST, "numeric except unsigned_long").and(
+            isType(percentile, dt -> dt.isNumeric() && dt != UNSIGNED_LONG, sourceText(), SECOND, "numeric except unsigned_long")
+        );
+    }
+
+    @Override
+    public boolean foldable() {
+        return field.foldable() && percentile.foldable();
+    }
+
+    public final Expression field() {
+        return field;
+    }
+
+    @Override
+    public DataType dataType() {
+        return field.dataType();
+    }
+
+    @Override
+    public final ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        var fieldEval = toEvaluator.apply(field);
+        var percentileEval = toEvaluator.apply(percentile);
+
+        return switch (PlannerUtils.toElementType(field.dataType())) {
+            case INT -> switch (PlannerUtils.toElementType(percentile.dataType())) {
```

Comment:
@not-napoleon has convinced me we should probably migrate from these kind of switch statements to a static `Map<List<DataType>, SomeClosure>` for this sort of thing. That way `resolveType` can check if the data type is in the `Map` and we don't have to maintain a sort of parallel infrastructure for functions.

Not sure it's worth doing now, but I think it's worth thinking about.

Reply:
I'll leave it out of this PR, as that idea looks nice, but there are edge cases that check other things apart from DataTypes (Like `SpatialCentroid`). So better to think it in parallel